<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>DADA2-Based Pipeline for Infering DNA Barcodes from Nanopore Sequence Data • DadaNanoBC</title>
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="DADA2-Based Pipeline for Infering DNA Barcodes from Nanopore Sequence Data">
<meta name="description" content="This package provides functions and a targets pipeline workflow to infer specimen barcodes from Nanopore sequencing data with multiplexed samples. It includes functions for automatic taxonomic assignments and taxonomy-based recognition of contaminants.">
<meta property="og:description" content="This package provides functions and a targets pipeline workflow to infer specimen barcodes from Nanopore sequencing data with multiplexed samples. It includes functions for automatic taxonomic assignments and taxonomy-based recognition of contaminants.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">DadaNanoBC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/basecalling.html">Basecalling</a></li>
    <li><a class="dropdown-item" href="articles/curation.html">Sequence curation</a></li>
    <li><a class="dropdown-item" href="articles/lab.html">Lab methods</a></li>
    <li><a class="dropdown-item" href="articles/primers.html">Primer design</a></li>
    <li><a class="dropdown-item" href="articles/workflow.html">Workflow steps</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="analysis-example.html">Example Report</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/markschl/DadaNanoBC/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="dadananobc-nanopore-barcoding-pipeline">DadaNanoBC: Nanopore barcoding pipeline<a class="anchor" aria-label="anchor" href="#dadananobc-nanopore-barcoding-pipeline"></a>
</h1></div>
<p>This R package provides functions for clustering and validating amplicon sequencing data from specimen barcoding projects using Oxford Nanopore. Although not tested, it should work for PacBio data as well.</p>
<ul>
<li>
<strong>sample demultiplexing</strong> of pooled reads</li>
<li>inferring <strong>barcode sequences</strong> using a <a href="https://benjjneb.github.io/dada2" class="external-link">DADA2</a>-based <a href="https://markschl.github.io/DadaNanoBC/articles/workflow.html#clustering">clustering strategy</a> to accurately resolve haplotypes/polymorphic sequences</li>
<li>automatic <strong>taxonomic assignents</strong> and recognition (down-ranking) of contaminant taxa</li>
<li>detailed output files (comprehensive Excel table for curation, HTML report, BAM alignment files)</li>
</ul>
<p><a href="https://markschl.github.io/DadaNanoBC/articles/workflow.html">See here for a comprehensive explanation</a></p>
<p>A fully-fledged <a href="https://books.ropensci.org/targets" class="external-link">{targets}</a> pipeline is provided for simple execution of a standard workflow (see <a href="#running-the-targets-pipeline">below</a>).</p>
<p><strong>Table of contents:</strong></p>
<ul>
<li><a href="#what-type-of-data-does-it-work-with">What type of data does it work with?</a></li>
<li><a href="#system-requirements">System requirements</a></li>
<li><a href="#preparation-lab-work-and-basecalling">Preparation: lab work and basecalling</a></li>
<li><a href="#installing-and-using-the-package">Installing and using the package</a></li>
<li><a href="#running-the-targets-pipeline">Running the {targets} pipeline</a></li>
<li><a href="#running-in-the-bash-console">Running in the bash console</a></li>
<li><a href="#required-software">Required software</a></li>
<li><a href="#useful-references">Useful references</a></li>
</ul>
<div class="section level2">
<h2 id="what-type-of-data-does-it-work-with">What type of data does it work with?<a class="anchor" aria-label="anchor" href="#what-type-of-data-does-it-work-with"></a>
</h2>
<ul>
<li>
<em>Basecalled</em> amplicon reads with sample-specific tags attached to both ends (dual-indexed)</li>
<li>Reads of sufficient quality (R10.4 chemistry)</li>
<li>Amplicon lengths of 0.5-1.5kb have been successfully tested</li>
<li>Works with fixed- and variable-length amplicons from different taxonomic groups that can be multiplexed and analyzed separately</li>
<li>different taxonomic database types can be auto-downloaded (more may be implemented in the future)</li>
</ul>
</div>
<div class="section level2">
<h2 id="system-requirements">System requirements<a class="anchor" aria-label="anchor" href="#system-requirements"></a>
</h2>
<ul>
<li>An UNIX environment (Linux, OS X or Windows with WSL2) with <em>Bash</em>
</li>
<li>An <a href="https://cran.r-project.org" class="external-link">R</a> installation</li>
<li>Four command-line programs (see <a href="#required-software">required software</a>); the package can always be installed without these, but error messages will appear when calling functions that rely on missing components.</li>
</ul>
<p>If <a href="https://nanoporetech.com/document/data-analysis#basecalling-overview" class="external-link">basecalling</a> of the Nanopore data has not yet been done: an Nvidia GPU with Cuda support (details <a href="https://markschl.github.io/DadaNanoBC/articles/basecalling.html">here</a>).</p>
</div>
<div class="section level2">
<h2 id="preparation-lab-work-and-basecalling">Preparation: lab work and basecalling<a class="anchor" aria-label="anchor" href="#preparation-lab-work-and-basecalling"></a>
</h2>
<ol style="list-style-type: decimal">
<li>
<a href="https://markschl.github.io/DadaNanoBC/articles/primers.html">Primer design</a> (<a href="https://markschl.github.io/DadaNanoBC/primer-design-example.html">example report</a>)</li>
<li><a href="https://markschl.github.io/DadaNanoBC/articles/lab.html">PCR, library preparation and sequencing</a></li>
<li><a href="https://markschl.github.io/DadaNanoBC/articles/basecalling.html">Basecalling</a></li>
</ol>
</div>
<div class="section level2">
<h2 id="installing-and-using-the-package">Installing and using the package<a class="anchor" aria-label="anchor" href="#installing-and-using-the-package"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"markschl/DadaNanoBC"</span><span class="op">)</span></span></code></pre></div>
<p>The package documentation is <a href="https://markschl.github.io/DadaNanoBC">found here</a>.</p>
<p>The package provides a mix of lower and higher level <a href="https://markschl.github.io/DadaNanoBC/reference/index.html">functions</a>, The core clustering function is <a href="https://markschl.github.io/DadaNanoBC/reference/infer_barcodes.html">infer_barcode</a> (detailed <a href="https://markschl.github.io/DadaNanoBC/articles/workflow.html#clustering">explanation here</a>). The high-level pipeline functions all have names starting with <code>do_</code> (e.g. <a href="https://markschl.github.io/DadaNanoBC/reference/do_trim_demux.html">do_trim_demux</a>).</p>
</div>
<div class="section level2">
<h2 id="running-the-targets-pipeline">Running the {targets} pipeline<a class="anchor" aria-label="anchor" href="#running-the-targets-pipeline"></a>
</h2>
<p>In the <code>R</code> console (for running in <em>Bash</em>, see <a href="#running-in-the-bash-console">below</a>):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"path/to/pipeline"</span><span class="op">)</span></span>
<span><span class="fu">DadaNanoBC</span><span class="fu">::</span><span class="fu"><a href="reference/init_pipeline.html">init_pipeline</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Some files are created, some need to be still placed</p>
<pre class="text"><code>path/to/pipeline/
 ├─ _targets.R               Pipeline R code
 └─ analysis/
    ├─ meta-ITS5-ITS4.xlsx   Metadata template (modify &amp; rename to 'meta.xlsx')
    ├─ config.yaml           Pipeline configuration (modify)
    └─ reads.fastq.gz    &lt;- Place base-called nanopore sequences here</code></pre>
<p>There may also be warnings about missing software, which have to be handled (see <a href="#required-software">required software</a>).</p>
<p>Next:</p>
<ul>
<li>Modify the example metadata file <code>meta-ITS5-ITS4.xlsx</code> to contain your primer sequences and sample metadata, and rename it to <strong>meta.xlsx</strong>
</li>
<li>Copy or move your <a href="https://markschl.github.io/DadaNanoBC/articles/basecalling.html">base-called</a> sequencing reads to <strong>reads.fastq.gz</strong>
</li>
<li>Adapt <strong>config.yaml</strong> to your needs; at the minimum, the <em>taxonony</em> section needs to contain the correct taxonomic database. &gt; Consider starting with a reduced read depth as in the example configuration (<code>max_sample_depth</code>/<code>max_contam_sample_depth</code>)</li>
</ul>
<p>Run the pipeline:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">targets</span><span class="fu">::</span><span class="fu">tar_make</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p><em>Optional</em>: Modify the number of parallel processes before calling <code>tar_make</code>: <code>Sys.setenv('DadaNanoBC_WORKERS' = 8)</code></p>
</blockquote>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>Some more files will appear in the analysis directory:</p>
<ul>
<li>
<strong>report.html</strong> (<a href="https://markschl.github.io/DadaNanoBC/analysis-example.html">example report</a>): detailed HTML report, useful for troubleshooting</li>
<li>
<strong>report.xlsx</strong>: Excel report with sequences and taxonomic assignments, which can be inspected for issues (see <a href="https://markschl.github.io/DadaNanoBC/articles/curation.html">curation</a>)</li>
<li>
<strong>separate_alignments</strong> (and/or <em>top_alignments</em>, <em>alignments</em> depending on configuration): aligned sequencing reads in BAM format, may be imported into sequence viewing/editing software such as Geneious/CLC Workbench/UGENE/IGV, etc.</li>
<li>
<strong>tmp</strong>: contains temporary (but sometimes useful) data such as the demultiplexed FASTQ sample files (sometimes <em>separate_alignments</em> with many small BAM files, depending on the configuration); if not needed you may delete the files</li>
</ul>
<blockquote>
<p>Inspect the HTML report for anomalies, maybe adjust some other settings then re-run with at higher depth (adjust/remove <code>max_sample_depth</code>/<code>max_contam_sample_depth</code>)</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="curation">Curation<a class="anchor" aria-label="anchor" href="#curation"></a>
</h3>
<p>Excel report contains sequences and issue flags, which might have to be addressed before exporting the sequences.</p>
<p><a href="https://markschl.github.io/DadaNanoBC/articles/curation.html">See curation tutorial</a></p>
</div>
<div class="section level3">
<h3 id="running-a-different-analysis">Running a different analysis<a class="anchor" aria-label="anchor" href="#running-a-different-analysis"></a>
</h3>
<p>Prepare another directory with an appropriate <code>meta.xlsx</code>, <code>reads.fastq.gz</code> and <code>config.yaml</code>. Then, set the path and run the analysis. Taxonomic databases and other files in the <em>taxdb</em> directory is reused.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span><span class="st">'DadaNanoBC_ANALYSIS_DIR'</span> <span class="op">=</span> <span class="st">'path/to/dir'</span><span class="op">)</span></span>
<span><span class="fu">targets</span><span class="fu">::</span><span class="fu">tar_make</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-in-the-bash-console">Running in the bash console<a class="anchor" aria-label="anchor" href="#running-in-the-bash-console"></a>
</h2>
<p>Install the package if not already done</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">'devtools::install_github("markschl/DadaNanoBC", upgrade = "never")'</span></span></code></pre></div>
<p>Initialize a directory where the pipeline is run</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> path/to/pipeline</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">'DadaNanoBC::init_pipeline(bash = TRUE)'</span></span></code></pre></div>
<p>Then make sure there are the files <code>meta.xlsx</code>, <code>reads.fastq.gz</code> and <code>config.yaml</code> present in the analysis directory as described <a href="#running-the-targets-pipeline">above</a>.</p>
<p>Run the pipeline (for the default <em>analysis</em> directory, may be <a href="#running-a-different-analysis">changed</a>):</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">./infer_barcodes</span> analysis</span></code></pre></div>
<p>Additional settings can be added, as in this more complicated/realistic example (see also <a href="#custom-program-location">custom program location</a>):</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">./infer_barcodes</span> WORKERS=8 vsearch=path/to/vsearch</span></code></pre></div>
<p><a href="#running-a-different-analysis">Different analysis</a> in another directory:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">./infer_barcodes</span> ANALYSIS_DIR=path/to/analysis</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="required-software">Required software<a class="anchor" aria-label="anchor" href="#required-software"></a>
</h2>
<p>The following tools are required:</p>
<ul>
<li>
<a href="https://github.com/markschl/seqtool" class="external-link">seqtool</a> for primer trimming/demultiplexing</li>
<li>
<a href="https://www.htslib.org" class="external-link">samtools</a> and <a href="https://github.com/lh3/minimap2" class="external-link">minimap2</a> for dealing with alignments</li>
<li>
<a href="https://github.com/torognes/vsearch" class="external-link">VSEARCH</a> for sequence-based taxonomic assignments</li>
</ul>
<p>The above call to <code><a href="reference/init_pipeline.html">init_pipeline()</a></code> checks for available software and gives the URLs of missing software. The following does the same:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DadaNanoBC</span><span class="fu">::</span><span class="fu"><a href="reference/check_system_requirements.html">check_system_requirements</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The message points to the download locations of missing tools. <em>Samtools</em>, <em>minimap2</em> and <em>VSEARCH</em> are available from different sources:</p>
<p><em>Debian/Ubuntu</em> (including Windows WSL2 with these OSes):</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">sudo</span> apt install <span class="at">-y</span> samtools minimap2 vsearch</span></code></pre></div>
<p><em>OS X with <a href="https://brew.sh" class="external-link">Homebrew</a>:</em></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">brew</span> install samtools minimap2 vsearch</span></code></pre></div>
<p><em>Conda package manager</em> (e.g. <a href="https://conda-forge.org/download" class="external-link">Conda-forge</a>) (run in the <a href="#running-in-the-bash-console">bash console</a>):</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-y</span> <span class="at">-n</span> DadaNanoBC <span class="at">-c</span> bioconda samtools minimap2 vsearch</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="ex">conda</span> activate DadaNanoBC</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="ex">./infer_barcodes</span> analysis_dir <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span></span></code></pre></div>
<p>Instructions for downloading <em>Seqtool</em> are given on the <a href="https://github.com/markschl/seqtool/releases" class="external-link">Github site</a>.</p>
<div class="section level3">
<h3 id="custom-program-location">Custom program location<a class="anchor" aria-label="anchor" href="#custom-program-location"></a>
</h3>
<p>R needs to know about the tool’s locations. If not placed at a standard location, either the <code>PATH</code> environment variable needs adjustment, or the paths can be communicated to DadaNanoBC.</p>
<p>In R (example with VSEARCH):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DadaNanoBC</span><span class="fu">::</span><span class="fu"><a href="reference/set_global_opts.html">set_global_opts</a></span><span class="op">(</span>vsearch <span class="op">=</span> <span class="st">'~/Downloads/vsearch-2.30.2-linux-x86_64/bin/vsearch'</span><span class="op">)</span></span>
<span><span class="fu">tar_make</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>In the Bash console, simply add <code>program=...</code> after the <code>infer_barcodes</code> call:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="bu">cd</span> path/to/pipeline</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="ex">./infer_barcodes</span> analysis WORKERS=8 vsearch=~/Downloads/vsearch-2.30.2-linux-x86_64/bin/vsearch</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="useful-references">Useful references<a class="anchor" aria-label="anchor" href="#useful-references"></a>
</h2>
<ul>
<li><a href="https://doi.org/10.1007/978-1-0716-3581-0_14" class="external-link">Srivathsan &amp; Meier (2024)</a></li>
<li>
<a href="https://doi.org/10.1111/1755-0998.14028" class="external-link">Hebert et al. (2024)</a> barcoded up to 100k samples with a single MinION run.</li>
<li>
<a href="https://www.researchgate.net/publication/393048029_Approaching_Full-Scale_DNA_Barcoding_for_North_American_Macrofungi_Highlights_from_the_MycoMap_Network" class="external-link">ONT barcoding efforts by the MycoMap network</a> (see also <a href="https://dx.doi.org/10.17504/protocols.io.dm6gpbm88lzp/v4" class="external-link">methods description</a>) (software used: <a href="https://github.com/ksahlin/NGSpeciesID" class="external-link">NGSpeciesID</a>)</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/markschl/DadaNanoBC/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing DadaNanoBC</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Markus Schlegel <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-3453-4783" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Markus Schlegel.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
