<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Basecalling • DadaNanoBC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Basecalling">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DadaNanoBC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basecalling.html">Basecalling</a>
    </li>
    <li>
      <a href="../articles/curation.html">Sequence curation</a>
    </li>
    <li>
      <a href="../articles/lab.html">Lab methods</a>
    </li>
    <li>
      <a href="../articles/primers.html">Primer design</a>
    </li>
    <li>
      <a href="../articles/workflow.html">Workflow steps</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Basecalling</h1>
            
      

      <div class="hidden name"><code>basecalling.Rmd</code></div>

    </div>

    
    
<p>If the computer has a Nvidia GPU with Cuda support (or a connected <a href="https://egpu.io" class="external-link">eGPU</a>), then basecalling may be done locally.
Alternatively, basecalling may be done on a different computer, e.g. a
HPC server with a GPU node.</p>
<div class="section level2">
<h2 id="model">Model<a class="anchor" aria-label="anchor" href="#model"></a>
</h2>
<p>Different <a href="https://software-docs.nanoporetech.com/dorado/latest/models/models" class="external-link">machine
learning models</a> are available for the basecalling, offering
different trade-offs between speed and accuracy. Our experience suggests
that the <code>sup</code> model gives substantially better results than
the <code>hac</code> model in terms of Q scores and recognized
primers/barcodes. Basecalling with <code>sup</code> may take from a few
hours to a full day for a MinION run, depending on the GPU setup.</p>
</div>
<div class="section level2">
<h2 id="on-your-computer">On your computer<a class="anchor" aria-label="anchor" href="#on-your-computer"></a>
</h2>
<p><a href="https://software-docs.nanoporetech.com/dorado/latest" class="external-link">Dorado</a>
is needed for this (<code>dorado</code> command should be available in
console). It’s installation size is quite large (a few Gigabytes).</p>
<p>The basecalling is done as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> path/to/DadaNanoBC</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># create the input directory (if not present)</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="va">run_name</span><span class="op">=</span>my_run</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> analysis/<span class="va">$run_name</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co"># Directory with the Nanopore data (may be on an external hard drive)</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="va">ont_data</span><span class="op">=</span>/media/user/run/data</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co"># The basecalling model</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="va">model</span><span class="op">=</span>sup</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co"># run the base-caller (may take a long time)</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co"># replace 'dorado' with full path to the 'dorado' binary if necessary</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="ex">dorado</span> basecaller <span class="va">$model</span> <span class="at">-r</span> <span class="at">--no-trim</span> <span class="at">--emit-fastq</span> <span class="va">$ont_data</span> <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-c</span> <span class="op">&gt;</span> analysis/<span class="va">$run_name</span>/reads.fastq.gz</span></code></pre></div>
<p>An easy way to find out the actual paths of the Nanopore data is to
drag &amp; drop the folders on the console, then copy the path
there.</p>
</div>
<div class="section level2">
<h2 id="on-a-hpc-server">On a HPC server<a class="anchor" aria-label="anchor" href="#on-a-hpc-server"></a>
</h2>
<p>Basecalling on a remote server is slightly more complicated. This
script shows how to do this on a SLURM server with a GPU node. The
<code>remote_dir</code> must contain a job submission script (based on
this <a href="https://github.com/markschl/DadaNanoBC/templates/base_calling/slurm_submit_basecaller.sh" class="external-link">template</a>);
it can be copied to the remote directory e.g. using an FTP client;
adjust the <code>dorado</code> path and your email address inside the
script.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="bu">cd</span> path/to/ont-barcoding</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co"># run directory</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="va">run_name</span><span class="op">=</span>my_run</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="va">run_dir</span><span class="op">=</span>analysis/<span class="va">$run_name</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co"># SSH login to HPC server</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="va">ssh</span><span class="op">=</span>user@server-address</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co"># Remote directory containing the submission script (see templates/slurm_submit_basecaller.sh)</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co"># The base-called data is placed in $remote_dir/$run_name</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="va">remote_dir</span><span class="op">=</span>projects/ONT</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co"># Nanopore data export directory</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="va">ont_data</span><span class="op">=</span>/media/user/run/data</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co"># The basecalling model</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="va">model</span><span class="op">=</span>sup</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co"># copy the data</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$run_dir</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="fu">rsync</span> <span class="at">-av</span> <span class="at">--info</span><span class="op">=</span>progress2 <span class="st">"</span><span class="va">$ont_data</span><span class="st">"</span> <span class="va">$ssh</span>:<span class="va">$remote_dir</span>/<span class="va">$run_name</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co"># submit the run</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="fu">ssh</span> <span class="va">$ssh</span> <span class="at">-t</span> <span class="st">"cd ~/</span><span class="va">$remote_dir</span><span class="st"> &amp;&amp; sbatch slurm_submit_basecaller.sh </span><span class="va">$model</span><span class="st"> </span><span class="va">$run_name</span><span class="st">"</span></span></code></pre></div>
<p>There should be a message: “submitted batch job …”. Wait for the
completion email and then execute:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">rsync</span> <span class="at">-av</span> <span class="va">$ssh</span>/<span class="va">$remote_dir</span>/<span class="va">$run_name</span>/output.fastq.gz <span class="va">$analysis</span>/<span class="va">$run_name</span>/reads.fastq.gz</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Markus Schlegel.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
