<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Sequence curation • DadaNanoBC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Sequence curation">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DadaNanoBC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basecalling.html">Basecalling</a>
    </li>
    <li>
      <a href="../articles/curation.html">Sequence curation</a>
    </li>
    <li>
      <a href="../articles/lab.html">Lab methods</a>
    </li>
    <li>
      <a href="../articles/primers.html">Primer design</a>
    </li>
    <li>
      <a href="../articles/workflow.html">Workflow steps</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Sequence curation</h1>
            
      

      <div class="hidden name"><code>curation.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="report">Report<a class="anchor" aria-label="anchor" href="#report"></a>
</h2>
<p>The results of the automated analysis are saved in an Excel document
named <strong>report.xlsx</strong>. There is one row per barcode, the
columns are explained by comments in the document.</p>
<p>The <strong>issues</strong> list should be investigated more closely,
although not all of them are critical. The list can be filtered.</p>
<div class="float">
<img src="img/report-overview.png" alt="report overview"><div class="figcaption">report overview</div>
</div>
<div class="section level3">
<h3 id="final-sequence-set">Final sequence set<a class="anchor" aria-label="anchor" href="#final-sequence-set"></a>
</h3>
<p>The <em>seq1</em>, [<em>seq2</em>, etc.] columns in the
<strong>Curation</strong> section contain the final sequences. In case
of problems, they can be edited. Changed sequences appear in <span style="color:blue">blue</span></p>
<blockquote>
<p>FASTA sequences can be exported by a simple copy&amp;paste from the
<em>FA1</em> [<em>FA2</em>, etc.] columns to a text file (works well
with LibreOffice).</p>
</blockquote>
<blockquote>
<p><strong>Excel:</strong> copy&amp;paste to a <em>Word document</em>,
then copy&amp;paste to the final text file or search field, etc.
<em>OR</em>: copy directly to a text file, then remove all double quotes
(“) with search&amp;replace.</p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="common-problems">Common problems<a class="anchor" aria-label="anchor" href="#common-problems"></a>
</h2>
<div class="section level3">
<h3 id="inconsistent-taxonomy">Inconsistent taxonomy<a class="anchor" aria-label="anchor" href="#inconsistent-taxonomy"></a>
</h3>
<p>The <strong>tax-mimatch</strong> issue indicates a mismatch between
the taxonomic name in the <em>morpho-taxon</em> entry and the
sequence-based identification. Possible reasons:</p>
<ul>
<li>mis-identified specimen</li>
<li>something got mixed up when preparing
specimens/extracting/amplifying</li>
<li>the automatic taxonomy assignment of barcode sequences did not work
correctly, or the taxonomic lineages obtained from GBIF are not
correct</li>
<li>contamination not automatically recognized (<a href="analysis/workflow.html#taxonomic-assignments">how it works</a>)</li>
</ul>
<p>It is advisable to check the <em>details</em> worksheet: Copy the
sample barcode combination (e.g. <em>bc018-bc183</em>) and filter the
<em>details</em> table to list <em>all</em> sequences from this sample.
The <em>group</em> column indicates the taxon cluster (the top being the
one shown in the main table).</p>
<blockquote>
<p>If the top taxon group is not the correct one, copy from the
“sequence” field(s) to “seq1” (“seq2”, …) in the main table (paste
values).</p>
</blockquote>
<blockquote>
<p>The <strong>contamination</strong> issue indicates that contamination
is very likely present and was automatically down-ranked. It may also be
good to investigate these samples more precisely in the <em>details</em>
tab.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="ambiguities">Ambiguities<a class="anchor" aria-label="anchor" href="#ambiguities"></a>
</h3>
<p>If the alignment consensus has ambiguous positions
(<em>ambig-consensus</em> issue), these may be inspected by clicking on
<em>→ data</em> to open a folder with alignment files. Use a sequence
viewer of your choice, e.g. Geneious, CLC Workbench or UGENE.</p>
<p>The <em>…seq_comparison.fasta</em> file may be inspected first.
Inconsistencies should be readily visible (example alignment view in
Geneious).</p>
<p><img src="img/seq-comparison.png"></p>
<p>The BAM file (and the corresponding reference consensus sequences)
offer additional insights:</p>
<p><img src="img/aln-detail.png"></p>
<p>In this case, a long homopolymer stretch results in an unclear
sequence.</p>
<blockquote>
<p>Consider copying the “group1_seq1_dominant_seq” (= most abundant
sequence in sample) to the <em>seq1</em> column in the <em>Curation</em>
section (without gaps).</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="list-of-issues">List of issues<a class="anchor" aria-label="anchor" href="#list-of-issues"></a>
</h3>
<ul>
<li><p><strong>ambig-consensus:</strong> At least one of the top taxon
consensus sequences have ambiguities, suggesting either unresolved
sequence variation, or sequencing errors e.g. in long homopolmyer
regions. To investigate, click on “→ data” to open a folder with
alignment files. (see also <a href="https://github.com/markschl/DadaNanoBC/blob/main/templates/analysis/config.commented.yaml" class="external-link"><em>consensus_threshold</em>
setting</a>)</p></li>
<li><p><strong>tax-mismatch:</strong> Inconsistency found between the
name in the <em>morpho-taxon</em> column and the <em>auto-assigned
taxon</em> (see also “matching ranks” column). Strong inconsistencies
might indicate a preparation error or contamination, although
contamination is often automatically recognized.</p></li>
</ul>
<blockquote>
<p>Note that the automatic taxonomy is not always correct; it depends on
the reference database and the <a href="https://github.com/markschl/DadaNanoBC/blob/main/templates/analysis/config.commented.yaml" class="external-link">confidence
threshold</a></p>
</blockquote>
<ul>
<li><p><strong>contamination:</strong> The most abundant taxon is
suspected to be a contaminating sequence and was therefore ignored. This
happens if the prepared specimen was not “pure”, and DNA from the host
or another specie was co-amplified. The top taxon is still listed in the
“details” sheet, and manual inspection of at least part of the putative
contaminants is recommended to ensure that there are no errors in the
automatic ranking (see also <a href="https://github.com/markschl/DadaNanoBC/blob/main/templates/analysis/config.commented.yaml" class="external-link"><em>contam_rank_delta</em>
setting</a>)</p></li>
<li><p><strong>known-seq-diffs:</strong> There is at least one mismatch
between the top taxon and the provided “known” sequence. To investigate,
click on “→ data” to open a folder containing
…seq_comparison.fasta</p></li>
<li><p><strong>known-seq-contamination:</strong> The provided “known”
sequence appears to be a contamination itself</p></li>
<li><p><strong>many-variants:</strong> Strong sequence variability for
the top taxon (&gt;4 abundant variants above the frequency threshold).
Maybe multiple individuals are present in the DNA mix?</p></li>
<li><p><strong>low-coverage:</strong> Only few Nanopore sequences
support the top taxon. Low-coverage samples may have more ambiguities,
and there is a higher risk for errors (possibly investigate the BAM
alignments) (see also <a href="https://github.com/markschl/DadaNanoBC/blob/main/templates/analysis/config.commented.yaml" class="external-link"><em>low_abund_threshold</em>
setting</a>)</p></li>
<li><p><strong>&lt; 3 times overabundant</strong>: The top taxon is less
than 3x more abundant than the second taxon. It may be worth to check
for contamination.</p></li>
<li><p><strong>Related taxa in mix</strong>: The taxonomy assignment
suggests that there are several related taxa present in the same sample.
It may be worth to check for contamination to be sure that the top taxon
is actually the one you are looking for.</p></li>
<li><p><strong>Re-mapping to consensus gives another consensus</strong>:
During the clustering workflow, consensus sequences were not identical
before and after a re-mapping of reads. Consider inspecting the
alignments (see <em>ambig-consensus</em>).</p></li>
<li><p><strong>[consensus-diffs]:</strong> At least one of the top taxon
consensus sequences does not match the <em>dominant sequence</em>. This
is not necessarily an issue (as the consensus is always reported) but is
an indication for noisy data and/or low-depth samples.</p></li>
<li><p><strong>[homopolymer-fix]:</strong> At least one of the top taxon
consensus sequences contains one or several long homopolymer stretches
(≥ 6 bp), where the consensus was not unambiguous, since Nanopore has
difficulties in determining the correct repeat length. Instead of the
consensensus, the sequence was taken from the prevalent sequencing
read.</p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Markus Schlegel.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
