<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Workflow steps • DadaNanoBC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Workflow steps">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DadaNanoBC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basecalling.html">Basecalling</a>
    </li>
    <li>
      <a href="../articles/curation.html">Sequence curation</a>
    </li>
    <li>
      <a href="../articles/lab.html">Lab methods</a>
    </li>
    <li>
      <a href="../articles/primers.html">Primer design</a>
    </li>
    <li>
      <a href="../articles/workflow.html">Workflow steps</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Workflow steps</h1>
            
      

      <div class="hidden name"><code>workflow.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="pipeline-steps">Pipeline steps<a class="anchor" aria-label="anchor" href="#pipeline-steps"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Search the primers and short sample indexes (located up-/downstream
of the primer sequences) to group them by sample (de-multiplex); remove
low-quality reads</li>
<li>Infer the barcode sequences using a sequential procedure based on <a href="https://benjjneb.github.io/dada2" class="external-link">DADA2</a> and/or fixed-threshold
clustering.</li>
<li>Auto-assign the taxonomy and compare with the name given by
morphological identification (if present) to validate and check for
contamination.</li>
<li>Compare with already known sequences (if present)</li>
<li>Export summary table, which can be further manually curated</li>
</ol>
</div>
<div class="section level2">
<h2 id="primer-search">Primer search<a class="anchor" aria-label="anchor" href="#primer-search"></a>
</h2>
<p>It is assumed that sample-specific tags are attached to primers on
both sides (dual indexing):</p>
<p><span style="color: #2a90b8ff">TAG1</span><span style="color: #b82ab3ff">FPRIMER</span>BARCODESEQUENCE<span style="color: #b82ab3ff">RPRIMER</span><span style="color: #2a90b8ff">TAG2</span></p>
<p><strong>Combining multiple amplicons</strong> is possible. Primers
are searched in the order that amplicons appear in the sample sheet.
Therefore, if there are <em>nested amplicons</em>, the shorter one
should be placed <em>at the end</em> in the sheet.</p>
</div>
<div class="section level2">
<h2 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<p>The basic procedure for the clustering is implemented by the <a href="https://markschl.github.io/DadaNanoBC/reference/infer_barcodes.html" class="external-link">infer_barcodes</a>
function. The procedure starts with DADA2 clustering, followed by
consensus building. Additional steps may follow depending on whether the
consensus is unambiguous (each alignment column supported by enough
identical bases), or not. The workflow is illustrated in this flowchart
Relevant configuration options are shown in <em>grey</em>.</p>
<p><img src="img/cluster-flowchart.png"></p>
<div class="section level3">
<h3 id="stepsstrategies">Steps/strategies<a class="anchor" aria-label="anchor" href="#stepsstrategies"></a>
</h3>
<div class="section level4">
<h4 id="dada2-denoising">DADA2 denoising<a class="anchor" aria-label="anchor" href="#dada2-denoising"></a>
</h4>
<p><a href="https://benjjneb.github.io/dada2" class="external-link">DADA2</a> is a popular
program for inferring sequence variants (ASV) in amplicon sequence data.
The program requires that there are least a few error-free reads
present. If there is not enough replication, fixed-threshold clustering
(see <a href="#fixed-threshold-clustering">below</a>) is applied
instead.</p>
<p><img src="img/cluster-dada.png"></p>
<p>The next step is to infer a <a href="#how-is-the-consensus-sequence-determined-">consensus sequence</a>
for the ASV, which should usually be identical with the ASV sequence.
Ambiguous bases in the consensus sequence are an indication of
unresolved sequence variation, which we further address by attempging <a href="#haplotype-splitting">haplotype splitting</a> or by more sensitive
denoising (see <a href="#flowchart">flowchart</a>).</p>
</div>
<div class="section level4">
<h4 id="haplotype-splitting">Haplotype splitting<a class="anchor" aria-label="anchor" href="#haplotype-splitting"></a>
</h4>
<p>For DADA2 ASVs with an <span style="color:00897c">ambiguous</span>
consensus, it is checked if there are <strong>two dominant
sequences</strong> (supported by a few identical sequences) serving as
references for splitting the sequencing reads into two similarly
abundant sub-clusters with (ideally) <em>unambiguous</em> sequences. The
splitting is done if the total number of ambiguous bases in the two
resulting consensus sequences is smaller than the number of ambiguous
bases in the original consensus of all combined reads (see also
non-public function <a href="https://markschl.github.io/DadaNanoBC/reference/try_split_haplotypes.html" class="external-link">try_split_haplotypes</a>).</p>
<p>This simple procedure is often successful and avoids the need for
repeating the DADA2 denoising with higher sensitivity, or it may help in
resolve InDel variation that might not be discovered by DADA2.</p>
<p><img src="img/split-haplo.png"></p>
</div>
<div class="section level4">
<h4 id="fixed-threshold-clustering">Fixed-threshold clustering<a class="anchor" aria-label="anchor" href="#fixed-threshold-clustering"></a>
</h4>
<p>Fixed-threshold clustering is applied for low-coverage samples that
don’t have enough sequence duplication (see
<code>dada_min_identical</code> and <code>dada_min_n0</code> settings in
<a href="https://markschl.github.io/DadaNanoBC/reference/infer_barcodes.html" class="external-link">infer_barcodes</a>).</p>
<p>By default, single-linkage is applied with a 97% identity threshold.
This may result in an <span style="color:00897c">ambiguous</span>
consensus sequence, which can still be manually inspected.</p>
<p><img src="img/cluster-fixed.png"></p>
</div>
</div>
<div class="section level3">
<h3 id="how-is-the-consensus-sequence-determined">How is the consensus sequence determined?<a class="anchor" aria-label="anchor" href="#how-is-the-consensus-sequence-determined"></a>
</h3>
<p>After DADA2 denoising, haplotype splitting and fixed-threshold
clustering, the raw (quality-filtered) reads are always aligned to the
representative sequence (ASV or dominant unique sequence) with <a href="https://lh3.github.io/minimap2" class="external-link">Minimap2</a>. The consensus
sequence is then obtained with <a href="https://www.htslib.org/doc/samtools-consensus.html" class="external-link">samtools
consensus</a>. With the default settings in <code>infer_barcodes</code>,
each base in the consensus sequence needs to be supported by at least
65% of the reads (weighted by quality scores) to be
<em>unambiguous</em>. Otherwise, an <a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation" class="external-link">ambiguous
base</a> is shown, which may indicate unresolved sequence variation.</p>
<p>The length of <strong>long homopolymer</strong> repeats is often not
very consistent between different sequencing reads, which often results
in <code>N</code> bases in the consensus. A simple frequency-based
approach is used to “fix” this situation: the homopolymer length is
taken from the <em>most abundant sequence</em> (see also non-public
function <a href="https://markschl.github.io/DadaNanoBC/reference/fix_homopolymers.html" class="external-link">fix_homopolymers</a>).</p>
<p>The correctness of the consensus is further ensured by re-mapping the
reads against the consensus sequence. If necessary, this step is
repeated until <em>samtools consensus</em> returns the same sequence in
two successive rounds.</p>
<p>The <em>final reported sequence</em> is always the consensus sequence
(see also <code><a href="../articles/curation.html">vignette("curation")</a></code>). If the barcode sample has
a sufficient read depth and sequencing quality, it should usually be
unambiguous and correct. For low-depth samples with ambiguities,
alignments might be inspected manually.</p>
</div>
<div class="section level3">
<h3 id="taxa-grouping">Taxa grouping<a class="anchor" aria-label="anchor" href="#taxa-grouping"></a>
</h3>
<p>Sequence variation <em>within</em> taxa (e.g. haplotypes) is
distinguished from <em>between-taxa</em> variation with simple
fixed-threshold single-linkage clustering done with <a href="https://www2.decipher.codes/Clusterize.html" class="external-link">DECIPHER</a>
(default: 97% threshold). This usually works well at least with the ITS
barcode.</p>
<blockquote>
<p>The most problematic case would be if there is a closely related
<strong>contaminant</strong> taxon in the DNA extract, which may not be
well distinghished from the target taxon.</p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="taxonomic-assignments">Taxonomic assignments<a class="anchor" aria-label="anchor" href="#taxonomic-assignments"></a>
</h2>
<p>Taxonomic names are automatically assigned with the the <a href="https://doi.org/10.1101/074161" class="external-link">SINTAX algorithm</a> using a
reference database (see <a href="https://markschl.github.io/DadaNanoBC/reference/infer_barcodes.html" class="external-link">do_assign_compare_taxonomy</a>).</p>
<p>Second, taxonomic names associated with the specimens are compared to
the sequence-based taxonomic assignments. From this, a <strong>taxonomic
overlap</strong> is calculated, which is the <em>proportion of matching
ranks in the taxonomic lineages</em> (as far as names are defined at
these levels). The lineages are retrieved from the <a href="https://doi.org/10.15468/39omei" class="external-link">GBIF backbone</a> taxonomy.</p>
<p>Inconsistent taxonomic labels may indicate errors during the
preparation or identification (or possibly the sequence-based
identification), or the presence of contamination. Potential
<strong>dominant contaminants</strong> are automatically flagged (see
<em>details</em> section in <a href="https://markschl.github.io/DadaNanoBC/reference/infer_barcodes.html" class="external-link">do_assign_compare_taxonomy
help page</a>).</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Markus Schlegel.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
